apps/language_server/test/experimental/provider/code_action/replace_with_underscore_test.exs

```diff
@@ -95,31 +111,12 @@ defmodule ElixirLS.LanguageServer.Experimental.Provider.CodeAction.ReplaceWithUn
         ]
       |> code_action("/project/file.ex", 0, "unused")
 
-    assert [] = apply(action)
-  end
-
-  test "handles nil context" do
-    assert {_, action} = code_action("other_var = 6", "/project/file.ex", 1, "not_found")
-
-    action = put_in(action, [:context], nil)
+    source_file = action.source_file
+    diagnostics = get_in(action, [:context, :diagnostics])
 
-    assert [] = apply(action)
-  end
-
-  test "handles nil diagnostics" do
-    assert {_, action} = code_action("other_var = 6", "/project/file.ex", 1, "not_found")
-
-    action = put_in(action, [:context, :diagnostics], nil)
-
-    assert [] = apply(action)
-  end
-
-  test "handles empty diagnostics" do
-    assert {_, action} = code_action("other_var = 6", "/project/file.ex", 1, "not_found")
-
-    action = put_in(action, [:context, :diagnostics], [])
-
-    assert [] = apply(action)
+    Enum.each(diagnostics, fn diagnostic ->
+      assert [] = ReplaceWithUnderscore.apply(source_file, diagnostic)
+    end)
   end
```

Comment:
why did you delete these tests? They were added for a reason (editors sometimes do this)

Reply:
I think I should have created a new test module `Provider.CodeActionTest` and put them there. They just were not applicable for `ReplaceWithUnderscore.apply` since we don't pass the whole action anymore (I decided to extract the file and diagnostics before calling `apply` since we need them in both code actions)

Reply:
> I think I should have created a new test module Provider.Handlers.CodeActionTest and put them there.

Agreed. 

> I decided to extract the file and diagnostics 

I think that change has made things a lot more difficult. why not just let the code actions get the request and return a list of code actions that they fine (or an empty list if they don't find anything?

This is what I'm talking about: https://github.com/elixir-lsp/elixir-ls/pull/776/files#r1131330271

Reply:
> there is no point in creating Provider.Handlers.CodeActionTest, isn't it?

That's the idea. Code that's so simple, it can't be wrong.